R/expand_model_out_val_grid.R

```diff
@@ -0,0 +1,107 @@
+#' Create expanded grid of valid task ID and output type value combinations
+#'
+#' @param config_tasks a list version of the content's of a hub's `tasks.json`
+#' config file, accessed through the `"config_tasks"` attribute of a `<hub_connection>`
+#' object or function [read_config()].
+#' @param round_id Character string. Round identifier. If the round is set to
+#' `round_id_from_variable: true`, IDs are values of the task ID defined in
+#' `round_id`. Otherwise should match config `round_id` values. Ignored if hub
+#' contains only a single round.
+#' @param required_only Logical. Whether to return combinations of required values only.
```

Comment:
I'd suggest making this parameter more specific, or at least making the documentation more specific, e.g. for which parameters is it looking for required/optional? based on what is in `return` I think this could say "Whether to return only combinations of required task ID variables and related output type IDs." 

Somewhat separately, I note that the name structure doesn't align syntactically with `include_opt_cols` for `create_model_out_submit_tmpl()`, as then if you are passing arguments to that function you are saying
"include optional columns"
and I would expect this option to be more of 
"include required values" or `include_req_vals` or something similar. So a not strongly held suggestion would be to change the name of this argument

Reply:
Addressed point 1 in e8c70dbda1c89241ea6cbeb592372689db2c7f3d

Reply:
What about if we changed `require_only` to `include_opt_vals`?

Reply:
That would completely change what the default behaviour of the function is assumed to be (i.e. default behaviour is that only required values are returned and you have to use the value of an argument to override that). That does not seem natural or intuitive at all to me. I definitely prefer `require_only`.

Reply:
For me there is still some semantic inconsistency. In my mind, the arguments `required_only` and `include_opt_cols` are doing similar things from a user perspective: they are allowing you to specify inclusion of optional/required elements in the dataframe. Therefore, the arguments should look more similar and be structured similarly.

Reply:
@LucieContamin, @rborchering @kathsherratt do you have any feelings on this? I think the most important is that arguments resonate naturally with the users so any input you have on a preference or other suggestions greatly appreciated.

Reply:
What about just minor changes to the names of each option, something like:
 - `required_only` --> `required_vals_only` (to increase specificity of what is required)
 - `include_opt_cols` --> `include_all_cols` (since the columns themselves aren't really optional, it's the values in them that are optional)

Reply:
I am not sure I totally understand the "optional" columns. All the columns should be present but the content of the column might be optional, for example: `location`: all the possible variables are optional but you have to have at least one variable included in the submission file. 
I cannot think of a situation where we have an "optional" columns that can be included or not in the projection file the team submitted.

Reply:
Hey @LucieContamin indeed on submission, all columns (including ones that have all optional values) must be completed.  The thing we are trying to lock down is what to do with columns that have all optional values when a template of only required values is generated. The default is to drop such columns when creating templates of required values only. But the current `include_opt_cols` (or perhaps even better @nickreich 's suggestion of `include_all_cols` allows us to include such columns as all NAs which some users might prefer as it shows all columns that need to be submitted and the `NA` columns added also are already the correct data type. As mentioned this is not the default behaviour but rather an option that I thought might be useful to some users.

Reply:
Clearly though the concept of `opt_cols` is causing some confusion and does not appear intuitive

Reply:
Thanks for the explanation, that's helpful. 
Personally, I will prefer a default behavior with all the columns, and have NAs as values for the ones with optional values. As the columns are required, I will expect the template with "required" only output to still have it, to emphasize that this information is required even if all possibles associated values are noted as "optional". 
Also, why would you need a template without all the columns? I cannot think of a situation where we need it.

Reply:
Thanks @LucieContamin that's really useful insight! We can certainly make the default behaviour  of `create_model_out_tmpl()` be that. At the minute it's wrapping the `expand_model_out_val_grid()` function which will be used in validation and there we do want columns with all optional values to be dropped so as to only validate required values. But we can certainly change the default value in `create_model_out_tmpl()`.

Re: the name of the argument though, do you prefer `include_opt_cols` or `include_all_cols`?

Reply:
I have a question about the `expand_model_out_val_grid()` behavior. I understand why we are dropping optional values but I think we should not drop any columns. I don't know what is your workflow so I apologize if I am wrong but to validate the required values, you need to have all the columns, even columns like `location`, as the "output_type*" and "value" column are defined by these columns. 
What I mean is, if you are validating that each combination of `target`, `horizon`, `location`, `origin_date` and `age group` contain all the required output type and output value in the expected format. You need to have the `location` information. Does that make sense?

Re: If we change the default behavior to include all columns and have the option to remove empty columns, I will just call it "remove_empty_cols".

Reply:
Re name of argument, `remove_empty_cols` sounds perfect and even more descriptive than `include_all_cols` if we agree to swap the function default.

Re: validation, this will happen in two steps. 
1. **Ensure all combinations of required values present:**
    When we are validating required values, we are only interested in the columns that actually have required values.  So we will use the output of `expand_model_out_val_grid(..., required_vals_only = TRUE)` to ensure that the submission file contains all required combinations across appropriate columns.
2. **Ensure submission values represent valid combinations of accepted values (optional + required)**
    This check will now compare the submission file to the full combinations of values and ensure the file either matches the output of `expand_model_out_val_grid()` or is a subset of it. In this check, all columns are involved (as optional values are included in the expanded grid) and we also ensure all combinations of values in the submission file are unique at this stage.

Having said that,  I will likely include a simpler validation test before these to check whether all expected columns for the round are present anyways.

Does that make sense?

Reply:
To summarise the suggestions so far:

1. Change default behaviour of `create_model_out_tmpl()` to include empty columns of optional values by default.
2. Use `remove_empty_cols` as the argument name to control above behaviour.
3. Change argument `required_only` to `required_vals_only` so as to be more descriptive.

Sound good @nickreich ?

Reply:
> 

Thanks for the details on the validation, it makes lot of sense and sounds good to me! Thank you very much again!

Reply:
Yes, this sounds good to me too! Thanks to both of you for working through this carefully.

Reply:
Implemented in 99d02c51669ac7b20475c81228cb0d39cde16606 & 62288557985732dcb9e6b1118f3f218be50b09b2